File: src/lib/api/git/staging.ts
// Index writes and the per-path working-tree operations beside them: stage and
// unstage whole files, folder roll-ups, hunks and lines, discard, ignore,
// untrack, and restore-from-commit. Mirrors `commands/staging.rs`.

import { invoke } from "@/lib/api/invoke";
import { applyLineRequestSchema, discardFilePreviewSchema } from "@/lib/api/schemas";
import { parse } from "@/lib/api/validate";
import { z } from "zod";
import type {
  ApplyLineRequest,
  DiscardFilePreview,
} from "./types";

export const stagingApi = {
  previewDiscardFile: async (
    path: string,
    file: string,
    previousFile: string | null,
    staged: boolean,
  ): Promise<DiscardFilePreview> =>
    parse(
      discardFilePreviewSchema,
      await invoke("preview_discard_file", { path, file, previousFile, staged }),
      "preview_discard_file",
    ),

  /** Stage one hunk from an unstaged diff, or unstage one hunk from a staged diff.
   * `expectedBody` is the displayed hunk's canonical body; the backend rejects the
   * stage if the current patch's hunk content no longer matches it. */
  applyHunk: async (
    path: string,
    file: string,
    staged: boolean,
    hunkIndex: number,
    expectedHeader: string,
    expectedBody: string,
  ) =>
    parse(
      z.string(),
      await invoke("apply_hunk", { path, file, staged, hunkIndex, expectedHeader, expectedBody }),
      "apply_hunk",
    ),

  /** Stage one changed line from an unstaged diff, or unstage one changed line from a staged diff. */
  applyLine: async (path: string, request: ApplyLineRequest) => {
    parse(applyLineRequestSchema, request, "apply_line");
    return parse(z.string(), await invoke("apply_line", { path, request }), "apply_line");
  },

  /** Stage several files atomically (one `git add -A`). */
  stageFiles: async (path: string, files: string[]) =>
    parse(z.string(), await invoke("stage_files", { path, files }), "stage_files"),

  /** Unstage several files atomically (one `git restore --staged`). */
  unstageFiles: async (path: string, files: string[]) =>
    parse(z.string(), await invoke("unstage_files", { path, files }), "unstage_files"),

  /** Discard one exact previewed file state. Staged changes restore from HEAD;
   * unstaged changes restore from the index. Renames carry `previousFile` so
   * both sides are handled as one logical change. */
  discardFile: async (
    path: string,
    file: string,
    previousFile: string | null,
    staged: boolean,
    expectedState: string,
  ) =>
    parse(
      z.string(),
      await invoke("discard_file", { path, file, previousFile, staged, expectedState }),
      "discard_file",
    ),

  /** Append one gitignore pattern to root `.gitignore`, or to `.git/info/exclude`
   * when `local` is true. */
  appendIgnorePattern: async (path: string, pattern: string, local: boolean) =>
    parse(
      z.string(),
      await invoke("append_ignore_pattern", { path, pattern, local }),
      "append_ignore_pattern",
    ),

  /** Drop a tracked path from the index while keeping the worktree leaf. */
  stopTracking: async (path: string, file: string) =>
    parse(z.string(), await invoke("stop_tracking", { path, file }), "stop_tracking"),

  /** ADR 0003: true when restoring `file` from `commitOid` would change on-disk bytes. */
  worktreeDiffersFromCommit: async (path: string, commitOid: string, file: string) =>
    parse(
      z.boolean(),
      await invoke("worktree_differs_from_commit", { path, commitOid, file }),
      "worktree_differs_from_commit",
    ),

  /** ADR 0003: true when `file` has a restorable (non-gitlink) blob at `commitOid`. */
  commitPathIsRestorable: async (path: string, commitOid: string, file: string) =>
    parse(
      z.boolean(),
      await invoke("commit_path_is_restorable", { path, commitOid, file }),
      "commit_path_is_restorable",
    ),

  /** ADR 0003: restore one path into the worktree from a commit (does not stage). */
  restorePathFromCommit: async (path: string, commitOid: string, file: string) =>
    parse(
      z.string(),
      await invoke("restore_path_from_commit", { path, commitOid, file }),
      "restore_path_from_commit",
    ),

  stageAll: async (path: string) =>
    parse(z.string(), await invoke("stage_all", { path }), "stage_all"),

  unstageAll: async (path: string) =>
    parse(z.string(), await invoke("unstage_all", { path }), "unstage_all"),
};

File: src-tauri/src/commands/staging.rs
//! Stage/unstage (file, hunk, line), discard-file, ignore, and restore-from-commit.

use super::{blocking, CommandError};
use crate::git;
use crate::git::types::DiscardFilePreview;

#[tauri::command]
pub async fn apply_hunk(
    path: String,
    file: String,
    staged: bool,
    hunk_index: usize,
    expected_header: String,
    expected_body: String,
) -> Result<String, CommandError> {
    blocking(move || {
        git::write::patch_staging::apply_hunk(
            &path,
            &file,
            staged,
            hunk_index,
            &expected_header,
            &expected_body,
        )
    })
    .await
}

#[tauri::command]
pub async fn apply_line(
    path: String,
    request: git::types::ApplyLineRequest,
) -> Result<String, CommandError> {
    blocking(move || git::write::patch_staging::apply_line(&path, &request)).await
}

#[tauri::command]
pub async fn stage_files(path: String, files: Vec<String>) -> Result<String, CommandError> {
    blocking(move || git::write::staging::stage_files(&path, &files)).await
}

#[tauri::command]
pub async fn unstage_files(path: String, files: Vec<String>) -> Result<String, CommandError> {
    blocking(move || git::write::staging::unstage_files(&path, &files)).await
}

#[tauri::command]
pub async fn preview_discard_file(
    path: String,
    file: String,
    previous_file: Option<String>,
    staged: bool,
) -> Result<DiscardFilePreview, CommandError> {
    blocking(move || {
        git::write::discard_file::preview_discard_file(
            &path,
            &file,
            previous_file.as_deref(),
            staged,
        )
    })
    .await
}

#[tauri::command]
pub async fn discard_file(
    path: String,
    file: String,
    previous_file: Option<String>,
    staged: bool,
    expected_state: String,
) -> Result<String, CommandError> {
    blocking(move || {
        git::write::discard_file::discard_file(
            &path,
            &file,
            previous_file.as_deref(),
            staged,
            &expected_state,
        )
    })
    .await
}

#[tauri::command]
pub async fn append_ignore_pattern(
    path: String,
    pattern: String,
    local: bool,
) -> Result<String, CommandError> {
    blocking(move || git::write::ignore::append_ignore_pattern(&path, &pattern, local)).await
}

#[tauri::command]
pub async fn stop_tracking(path: String, file: String) -> Result<String, CommandError> {
    blocking(move || git::write::staging::stop_tracking(&path, &file)).await
}

/// ADR 0003: true when restoring `file` from `commit_oid` would change on-disk bytes.
#[tauri::command]
pub async fn worktree_differs_from_commit(
    path: String,
    commit_oid: String,
    file: String,
) -> Result<bool, CommandError> {
    blocking(move || {
        git::write::restore_path::worktree_differs_from_commit(&path, &commit_oid, &file)
    })
    .await
}

/// ADR 0003: true when `file` has a restorable (non-gitlink) blob at `commit_oid`.
/// Lets the merged-selection surface probe the selection-tip commit before
/// offering Restore for a union path.
#[tauri::command]
pub async fn commit_path_is_restorable(
    path: String,
    commit_oid: String,
    file: String,
) -> Result<bool, CommandError> {
    blocking(move || {
        Ok::<_, CommandError>(git::write::restore_path::commit_path_is_restorable(
            &path,
            &commit_oid,
            &file,
        ))
    })
    .await
}

/// ADR 0003: restore one path into the worktree from a commit (does not stage).
#[tauri::command]
pub async fn restore_path_from_commit(
    path: String,
    commit_oid: String,
    file: String,
) -> Result<String, CommandError> {
    blocking(move || git::write::restore_path::restore_path_from_commit(&path, &commit_oid, &file))
        .await
}

#[tauri::command]
pub async fn stage_all(path: String) -> Result<String, CommandError> {
    blocking(move || git::write::staging::stage_all(&path)).await
}

#[tauri::command]
pub async fn unstage_all(path: String) -> Result<String, CommandError> {
    blocking(move || git::write::staging::unstage_all(&path)).await
}
